File: src/batch/runner.ts
// Unified batch orchestrator. Drives every FlowEvent through the decision
// engine / state machine, executes the decided action, generates LLM copy only
// after a decision, and appends every transition to the audit log.
// The engine decides; this loop only executes and logs.

import { FlowEvent, FlowType, AgentState, AuditEntry, RecoveryContext } from '../flows/types';
import { classify } from '../diagnosis/classify';
import { stepState } from '../decisions/stateMachine';
import { executeAction } from '../execution/flows';
import { generateCopy, generateExceptionSummary } from '../execution/copy';
import { AuditStore } from '../audit/store';
import { MAX_BATCH_ATTEMPTS } from '../decisions/rules';

export interface BatchResult {
  auditEntries: AuditEntry[];
  escapes: { eventId: string; subject: string; flow: FlowType }[];
  exceptionSummary: string;
  batchAttemptLimitHit: boolean;
}

export interface RunnerOpts {
  now?: number;
  maxBatchAttempts?: number;
}

// Subset of RecoveryContext built once per event (fields constant across attempts).
type RecoveryContextBase = Pick<
  RecoveryContext,
  'eventId' | 'flow' | 'customerId' | 'invoiceId' | 'amount' | 'currency' | 'reason' | 'now'
>;

export function runBatch(events: FlowEvent[], audit: AuditStore, opts?: RunnerOpts): BatchResult {
  let batchAttemptLimitHit = false;
  const maxBatch = opts?.maxBatchAttempts ?? MAX_BATCH_ATTEMPTS();
  let used = 0;

  // Per-customer touch tracking across events (blast radius, not per-event).
  const touchesByCustomer = new Map<string, number>();

  const escapes: BatchResult['escapes'] = [];
  const escalatedEvents: AuditEntry[] = [];

  for (const event of events) {
    if (used >= maxBatch) { batchAttemptLimitHit = true; break; }

    const reason = classify(event);
    const now = opts?.now ?? Date.now();

    const ctxBase: RecoveryContextBase = {
      eventId: event.eventId,
      flow: event.flow,
      customerId: event.customerId,
      invoiceId: event.invoiceId ?? '',
      amount: event.amount,
      currency: event.currency,
      reason,
      now,
    };

    let attempt = 0;
    let state: AgentState = 'detected';
    let guard = 0;

    while (!isTerminal(state) && guard < 100) {
      guard++;
      if (used >= maxBatch) { batchAttemptLimitHit = true; break; }
      used++;

      const ctx = flowContext(event, ctxBase, attempt, touchesByCustomer);
      const { next, audit: entry } = stepState(ctx);
      audit.append(entry);
      state = next.state;

      switch (next.decision) {
        case 'retry':
        case 'contact': {
          const res = executeAction(event, event.flow, reason, attempt);
          touchesByCustomer.set(event.customerId, (touchesByCustomer.get(event.customerId) ?? 0) + 1);
          const actionEntry: AuditEntry = {
            ...entry,
            ruleFired: res.recovered ? 'recovered' : entry.ruleFired,
            decision: res.recovered ? 'recover' : entry.decision,
            state: res.recovered ? 'recovered' : entry.state === 'contact_scheduled' ? 'contacting' : 'retrying',
            channel: res.channel,
            outcome: res.detail ?? '',
            actor: 'policy_engine',
            attempt,
            notes: res.note,
          };
          audit.append(actionEntry);
          attempt++;
          if (res.recovered) {
            audit.append({ ...actionEntry, ruleFired: 'recovered', decision: 'recover', state: 'recovered', outcome: `Recovered ${(res.recoveredAmount ?? 0) / 100} ${event.currency}` });
            state = 'recovered';
          }
          break;
        }
        case 'escalate': {
          // LLM copy ONLY after policy decided to escalate.
          const copy = generateCopy({
            eventId: event.eventId,
            flow: event.flow,
            customerName: event.customerName,
            customerEmail: event.customerEmail,
            amountInRupees: event.amount / 100,
            reason,
            invoiceId: event.invoiceId,
            overdueDays: (event.signal?.overdue_days as number) ?? undefined,
          });
          audit.append({
            ...entry,
            ruleFired: 'no_op',
            decision: 'escalate',
            actor: 'llm_copy',
            state: 'escalated',
            outcome: `Generated ${copy.channel} message: ${copy.subject}`,
          });
          escapes.push({ eventId: event.eventId, subject: copy.subject, flow: event.flow });
          if (!escalatedEvents.some((e) => e.eventId === event.eventId)) escalatedEvents.push(entry);
          state = 'escalated';
          break;
        }
        case 'hold': {
          // Park the event for a later scheduled run (mandate window, dispute,
          // PTP-wait). Not terminal, but we must NOT re-evaluate in this loop
          // (otherwise it would spin forever). A later scheduled run re-evaluates.
          audit.append({
            ...entry,
            ruleFired: entry.ruleFired,
            decision: 'hold',
            actor: 'policy_engine',
            state: 'waiting_on_ptp',
            outcome: `Held ${event.eventId} (${entry.ruleFired}) pending next scheduled window`,
          });
          state = 'waiting_on_ptp';
          break;
        }
        case 'none': {
          // No action warranted (e.g. receivable under dunning threshold).
          // Park it as a terminal no-op for this run; a later run re-evaluates.
          audit.append({
            ...entry,
            ruleFired: entry.ruleFired,
            decision: 'none',
            actor: 'policy_engine',
            state: 'abandoned',
            outcome: `No recovery action warranted for ${event.eventId} (${entry.ruleFired})`,
          });
          state = 'abandoned';
          break;
        }
        case 'suppress':
        case 'abandon':
        case 'recover':
        default:
          // terminal-ish; loop guard will exit
          break;
      }
    }
  }

  const exceptionSummary = generateExceptionSummary(
    escalatedEvents.map((e) => ({
      eventId: e.eventId,
      flow: e.flow,
      reason: e.reasonBucket,
      amountInRupees: (e.amount ?? 0) / 100,
    })),
  ).text;

  return { auditEntries: audit.all(), escapes, exceptionSummary, batchAttemptLimitHit };
}

function flowContext(
  event: FlowEvent,
  base: RecoveryContextBase,
  attempt: number,
  touchesByCustomer: Map<string, number>,
) {
  const s = event.signal ?? {};
  return {
    ...base,
    attempt,
    touchesForCustomer: touchesByCustomer.get(event.customerId) ?? 0,
    activePromiseToPayDate: (s.ptp_date as number) ?? null,
    overdueDays: (s.overdue_days as number) ?? 0,
    retryWindow: s.retry_window ? { start: (s.retry_window as any).start, end: (s.retry_window as any).end } : undefined,
    dncFlag: (s.dnc_flag as boolean) ?? false,
    cartValue: (s.cart_value as number) ?? 0,
    visits: (s.visits as number) ?? 1,
    rollingSuccessRate: (s.success_rate as number) ?? 1,
  };
}

function isTerminal(s: AgentState): boolean {
  return s === 'recovered' || s === 'escalated' || s === 'abandoned' || s === 'suppressed' || s === 'waiting_on_ptp';
}

import * as fs from 'fs';
import * as path from 'path';
export function loadEvents(dir = 'data/flows'): FlowEvent[] {
  const p = path.join(process.cwd(), dir);
  const files = fs.readdirSync(p).filter((f) => f.endsWith('.json'));
  return files.map((f) => JSON.parse(fs.readFileSync(path.join(p, f), 'utf8')) as FlowEvent);
}

File: go/rz/runner.go
package rz

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// Unified batch orchestrator. Drives every FlowEvent through the decision
// engine / state machine, executes the decided action, generates LLM copy only
// after a decision, and appends every transition to the audit log.
// The engine decides; this loop only executes and logs.

type Escape struct {
	EventID string
	Subject string
	Flow    FlowType
}

type BatchResult struct {
	AuditEntries         []AuditEntry
	Escapes              []Escape
	ExceptionSummary     string
	BatchAttemptLimitHit bool
}

type RunnerOpts struct {
	Now              int64
	MaxBatchAttempts *int
}

// RunBatch runs the whole batch into the audit store.
func RunBatch(events []*FlowEvent, audit *AuditStore, opts RunnerOpts) (BatchResult, error) {
	batchAttemptLimitHit := false
	maxBatch := MaxBatchAttempts()
	if opts.MaxBatchAttempts != nil {
		maxBatch = *opts.MaxBatchAttempts
	}
	used := 0

	touchesByCustomer := map[string]int{}
	escapes := []Escape{}
	escalatedEvents := []AuditEntry{}
	escalatedIDs := map[string]bool{}
	now := opts.Now

	for _, event := range events {
		// Idempotency gate: a duplicate eventId must never re-execute or
		// re-touch. Claim is atomic (serialized by the store mutex) and survives
		// restarts because it is hydrated from the existing log, so concurrent
		// or replayed webhook deliveries are at-most-once.
		firstSeen, err := audit.Claim(event.EventID)
		if err != nil {
			return BatchResult{}, err
		}
		if !firstSeen {
			dup := AuditEntry{
				EventID:      event.EventID,
				Timestamp:    Iso(now),
				Flow:         event.Flow,
				ReasonBucket: Classify(event),
				RuleFired:    RuleDuplicateSuppress,
				Decision:     DecisionSuppress,
				Actor:        ActorPolicyEngine,
				Outcome:      "Duplicate eventId already processed; suppressed by idempotency gate (at-most-once)",
				State:        AgentSuppressed,
				CustomerID:   event.CustomerID,
				InvoiceID:    event.InvoiceID,
				Amount:       i64p(event.Amount),
				Currency:     event.Currency,
			}
			if err := audit.Append(dup); err != nil {
				return BatchResult{}, err
			}
			continue
		}

		if used >= maxBatch {
			batchAttemptLimitHit = true
			break
		}

		reason := Classify(event)

		ctxBase := RecoveryContext{
			EventID:    event.EventID,
			Flow:       event.Flow,
			CustomerID: event.CustomerID,
			InvoiceID:  event.InvoiceID,
			Amount:     event.Amount,
			Currency:   event.Currency,
			Reason:     reason,
			Now:        now,
		}

		attempt := 0
		state := AgentState(AgentDetected)
		guard := 0

		for !runningTerminal(state) && guard < 100 {
			guard++
			if used >= maxBatch {
				batchAttemptLimitHit = true
				break
			}
			used++

			ctx := flowContext(event, ctxBase, attempt, touchesByCustomer)
			sr, err := StepState(ctx)
			if err != nil {
				return BatchResult{}, err
			}
			next, entry := sr.Next, sr.Audit
			if err := audit.Append(entry); err != nil {
				return BatchResult{}, err
			}
			state = next.State

			switch next.Decision {
			case DecisionRetry, DecisionContact:
				res := ExecuteAction(event, event.Flow, reason, attempt, ModeSandbox)
				touchesByCustomer[event.CustomerID]++
				actionEntry := entry
				actionEntry.RuleFired = RuleRecovered
				actionEntry.Decision = DecisionRecover
				if !res.Recovered {
					actionEntry.RuleFired = entry.RuleFired
					actionEntry.Decision = entry.Decision
				}
				actionEntry.State = AgentRecovered
				switch {
				case res.Recovered:
					actionEntry.State = AgentRecovered
				case entry.State == AgentContactScheduled:
					actionEntry.State = AgentContacting
				default:
					actionEntry.State = AgentRetrying
				}
				actionEntry.Channel = res.Channel
				actionEntry.Outcome = res.Detail
				actionEntry.Actor = ActorPolicyEngine
				actionEntry.Attempt = intp(attempt)
				actionEntry.Notes = res.Note
				if err := audit.Append(actionEntry); err != nil {
					return BatchResult{}, err
				}
				attempt++
				if res.Recovered {
					recEntry := actionEntry
					recEntry.RuleFired = RuleRecovered
					recEntry.Decision = DecisionRecover
					recEntry.State = AgentRecovered
					recAmt := float64(0)
					if res.RecoveredAmount > 0 {
						recAmt = float64(res.RecoveredAmount)
					}
					recEntry.Outcome = "Recovered " + formatJS(recAmt/100) + " " + event.Currency
					if err := audit.Append(recEntry); err != nil {
						return BatchResult{}, err
					}
					state = AgentRecovered
				}
			case DecisionEscalate:
				copy := GenerateCopy(CopyInput{
					EventID:        event.EventID,
					Flow:           event.Flow,
					CustomerName:   event.CustomerName,
					CustomerEmail:  event.CustomerEmail,
					AmountInRupees: float64(event.Amount) / 100,
					Reason:         reason,
					InvoiceID:      event.InvoiceID,
					OverdueDays:    int(numField(event.Signal, "overdue_days")),
				})
				llmEntry := entry
				llmEntry.RuleFired = RuleNoOp
				llmEntry.Decision = DecisionEscalate
				llmEntry.Actor = ActorLLMCopy
				llmEntry.State = AgentEscalated
				llmEntry.Outcome = "Generated " + copy.Channel + " message: " + copy.Subject
				if err := audit.Append(llmEntry); err != nil {
					return BatchResult{}, err
				}
				escapes = append(escapes, Escape{EventID: event.EventID, Subject: copy.Subject, Flow: event.Flow})
				if !escalatedIDs[event.EventID] {
					escalatedIDs[event.EventID] = true
					escalatedEvents = append(escalatedEvents, entry)
				}
				state = AgentEscalated
			case DecisionHold:
				holdEntry := entry
				holdEntry.RuleFired = entry.RuleFired
				holdEntry.Decision = DecisionHold
				holdEntry.Actor = ActorPolicyEngine
				holdEntry.State = AgentWaitingOnPtp
				holdEntry.Outcome = "Held " + event.EventID + " (" + string(entry.RuleFired) + ") pending next scheduled window"
				if err := audit.Append(holdEntry); err != nil {
					return BatchResult{}, err
				}
				state = AgentWaitingOnPtp
			case DecisionNone:
				noneEntry := entry
				noneEntry.RuleFired = entry.RuleFired
				noneEntry.Decision = DecisionNone
				noneEntry.Actor = ActorPolicyEngine
				noneEntry.State = AgentAbandoned
				noneEntry.Outcome = "No recovery action warranted for " + event.EventID + " (" + string(entry.RuleFired) + ")"
				if err := audit.Append(noneEntry); err != nil {
					return BatchResult{}, err
				}
				state = AgentAbandoned
			default:
				// suppress / abandon / recover / none: terminal-ish; loop guard exits
			}
		}
	}

	exInput := make([]ExceptionInput, 0, len(escalatedEvents))
	for _, e := range escalatedEvents {
		amt := float64(0)
		if e.Amount != nil {
			amt = float64(*e.Amount)
		}
		exInput = append(exInput, ExceptionInput{
			EventID:        e.EventID,
			Flow:           e.Flow,
			Reason:         e.ReasonBucket,
			AmountInRupees: amt / 100,
		})
	}
	allEntries, err := audit.All()
	if err != nil {
		return BatchResult{}, err
	}
	return BatchResult{
		AuditEntries:         allEntries,
		Escapes:              escapes,
		ExceptionSummary:     GenerateExceptionSummary(exInput).Text,
		BatchAttemptLimitHit: batchAttemptLimitHit,
	}, nil
}

func flowContext(event *FlowEvent, base RecoveryContext, attempt int, touchesByCustomer map[string]int) *RecoveryContext {
	s := event.Signal
	if s == nil {
		s = map[string]any{}
	}
	ctx := base
	ctx.Attempt = attempt
	ctx.TouchesForCustomer = touchesByCustomer[event.CustomerID]
	if v, ok := s["ptp_date"]; ok && v != nil {
		if n, ok2 := toInt64(v); ok2 {
			ctx.ActivePromiseToPayDate = &n
		}
	}
	ctx.OverdueDays = int(numField(s, "overdue_days"))
	if rw, ok := s["retry_window"].(map[string]any); ok {
		start, okS := toInt64(rw["start"])
		end, okE := toInt64(rw["end"])
		if okS && okE {
			w := &RetryWindow{Start: start, End: end}
			ctx.RetryWindow = w
		}
	}
	ctx.DNCFlag = boolField(s, "dnc_flag")
	ctx.CartValue = int64(numField(s, "cart_value"))
	ctx.Visits = int(numField(s, "visits"))
	ctx.RollingSuccessRate = numField(s, "success_rate")
	return &ctx
}

func toInt64(v any) (int64, bool) {
	switch n := v.(type) {
	case float64:
		return int64(n), true
	case int:
		return int64(n), true
	case int64:
		return n, true
	case float32:
		return int64(n), true
	}
	return 0, false
}

func runningTerminal(s AgentState) bool {
	switch s {
	case AgentRecovered, AgentEscalated, AgentAbandoned, AgentSuppressed, AgentWaitingOnPtp:
		return true
	}
	return false
}

// LoadEvents reads all normalized flow events from a directory (*.json), sorted
// by filename for cross-platform determinism.
func LoadEvents(dir string) ([]*FlowEvent, error) {
	if dir == "" {
		dir = filepath.Join("data", "flows")
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	var names []string
	for _, ent := range entries {
		if !ent.IsDir() && strings.HasSuffix(ent.Name(), ".json") {
			names = append(names, ent.Name())
		}
	}
	sort.Strings(names)
	var events []*FlowEvent
	for _, n := range names {
		b, err := os.ReadFile(filepath.Join(dir, n))
		if err != nil {
			return nil, err
		}
		var e FlowEvent
		if err := json.Unmarshal(b, &e); err != nil {
			return nil, err
		}
		events = append(events, &e)
	}
	return events, nil
}
